### PR TITLE
feat(plugins): add full conversation context to before_tool_call hook

### PR DESCRIPTION
**Pull request by Reapor-Yurnero**
## Summary

Describe the problem and fix in 2–5 bullets:

- Problem: `before_tool_call` only exposed `toolName` / `params`, which is not enough for an external policy / guardrail plugin to evaluate a tool call in context to look out for prompt injections / policy-violated actions etc.
- Why it matters: contextual inspection needs the same model-facing view that produced the tool call: effective system prompt, conversation snapshot, and available tool definitions.
- What changed: `before_tool_call` now optionally includes `systemPrompt`, `messages`, and `tools`, wired from the embedded runner into the existing hook payload. 
- What did NOT change (scope boundary): no new hook, no change to hook timing/order, no compatibility issue, no change to hook actions/returns, no bundled guardrail framework, no config changes.

## Change Type (select all)

- [x] Feature

## Scope (select all touched areas)

- [x] API / contracts

## Linked Issue/PR

- Related #6095 (that PR attempted to introduce a modular guardrails framework and got 50+ upvotes and many discussions; this PR splits out a minimal/narrow change that can enable guardrails plugins to inspect tool calls, as requested per discord discussions)
- Related #12617
- Related #48503

## User-visible / Behavior Changes

- Plugin authors can now read `systemPrompt`, `messages`, and `tools` from `before_tool_call`.
- Existing plugins remain compatible; the added fields are optional and additive.
- No user-facing behavior/config changes in the default runtime.

## Diagram (if applicable)

```text
Before:
[model emits tool call] 
  -> [before_tool_call gets toolName + params only]

After:
[model emits tool call]
  -> [before_tool_call gets toolName + params + systemPrompt + messages + tools]
  -> [plugin decides allow/block/param rewrite]
```

## Security Impact (required)

- New permissions/capabilities? (`Yes/No`) Yes
- Secrets/tokens handling changed? (`Yes/No`) No
- New/changed network calls? (`Yes/No`) No
- Command/tool execution surface changed? (`Yes/No`) No
- Data access scope changed? (`Yes/No`) Yes
- If any `Yes`, explain risk + mitigation:
  - This expands what a `before_tool_call` plugin can observe by adding context (`systemPrompt`, `messages`, `tools`).
  - The hook action space is unchanged: plugins can still only allow, block, or modify params as before.
  - The change is *additive* and limited to already-installed plugin code running on the same hook path.

### Environment

- OS: Ubuntu/Linux
- Runtime/container: local `pnpm` dev build
- Model/provider: `openai/gpt-5.4`
- Integration/channel (if any): none
- Relevant config (redacted): local standalone external plugin linked via `openclaw plugins install --link`, using `beforeToolCall` only

### Steps

1. Install a standalone external plugin that listens to `before_tool_call` and forwards the payload to a local dummy HTTP endpoint.
2. Start the gateway and run an agent prompt that causes a tool call.
3. Inspect the captured payload at the dummy endpoint.

### Expected

- The plugin receives `systemPrompt`, `messages`, and `tools` alongside the existing `toolName` / `params`.

### Actual

- Verified locally: the captured payload includes full conversation context, effective system prompt, and tool definitions before tool execution.

## Evidence

Attach at least one:

- [ ] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

End-to-end local test with an external standalone plugin posting hook payloads to a dummy HTTP serve

```text
Request 1
roles: system,user,assistant
tool_count: 27
metadata_keys: cygnal_bypass, openclaw_hook, openclaw_tool_call_id, openclaw_tool_name
last_tool_call.name: read
last_tool_call.arguments: {"path":"/home/autoswan/.openclaw/workspace/AGENTS.md","offset":1,"limit":1}

Request 2
roles: system,user,assistant,tool,assistant,user,assistant
tool_count: 27
metadata_keys: cygnal_bypass, openclaw_hook, openclaw_tool_call_id, openclaw_tool_name
last_tool_call.name: read
last_tool_call.arguments: {"path":"/home/autoswan/.openclaw/workspace/TOOLS.md","offset":1,"limit":1}
```

Gateway trace from the same run:

```text
[plugins] [hooks] running before_tool_call (1 handlers, sequential)
[plugins] stage=before_tool_call monitor:start url=http://127.0.0.1:18888/dummy
[plugins] stage=before_tool_call monitor:response-headers status=200
[plugins] stage=before_tool_call monitor:success status=200
```

## Automated Verification

- `pnpm build`, `pnpm test:contracts:plugins`, `pnpm check` passes
- `pnpm test` failed exactly the same as the upstream main current state.
- Ran `codex review --base upstream/main` and incorporated the suggestion to propagate before_tool_call context to bundled MCP/LSP tools so the hook behavior is consistent across various tool sources.

## Human Verification (required)

What you personally verified (not just CI), and how:

- Verified scenarios:
  - focused unit test for `before_tool_call` context forwarding
  - full local build
  - end-to-end local test with an external standalone plugin posting hook payloads to a dummy HTTP server
- Edge cases checked:
  - multi-step session where the second tool call payload included prior conversation/tool history
  - existing hook behavior remained unchanged apart from added optional fields
- What you did **not** verify:
  - non-embedded runner paths
  - actually published marketplace packaging/distribution flow

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

If a bot review conversation is addressed by this PR, resolve that conversation yourself. Do not leave bot review conversation cleanup for maintainers.

## Compatibility / Migration

- Backward compatible? (`Yes/No`) Yes
- Config/env changes? (`Yes/No`) No
- Migration needed? (`Yes/No`) No
- If yes, exact upgrade steps:

## Risks and Mitigations

- Risk: plugins on `before_tool_call` can now observe more runtime context than before.
  - Mitigation: the change is additive and optional only, and is therefore fully backward-compatible.

### CODE REVIEW

File: src/agents/pi-tools.before-tool-call.ts [hunk 1]
```diff
@@ -197,6 +205,9 @@ export async function runBeforeToolCallHook(args: {
         params: normalizedParams,
         ...(args.ctx?.runId && { runId: args.ctx.runId }),
         ...(args.toolCallId && { toolCallId: args.toolCallId }),
+        ...(args.ctx?.getSystemPrompt && { systemPrompt: args.ctx.getSystemPrompt() }),
+        ...(args.ctx?.getMessages && { messages: args.ctx.getMessages() }),
+        ...(args.ctx?.getTools && { tools: args.ctx.getTools() }),
```
[the remark this answers is not in the document]
Comment (Reapor-Yurnero) [reply src/agents/pi-tools.before-tool-call.ts]: it's okay as this field is additive and optional anyway

File: src/agents/pi-embedded-runner/run/attempt.ts [hunk 2]
```diff
@@ -839,6 +871,7 @@ export async function runEmbeddedAttempt(
         throw new Error("Embedded agent session missing");
       }
       const activeSession = session;
+      beforeToolCallMessagesProvider = () => activeSession.messages;
```
[the remark this answers is not in the document]
Comment (Reapor-Yurnero) [reply src/agents/pi-embedded-runner/run/attempt.ts:874]: this is superceded by the codex review below

File: src/agents/pi-tools.before-tool-call.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Reapor-Yurnero) [reply src/agents/pi-tools.before-tool-call.ts]: good catch! addressed in a new commit.

File: src/agents/pi-tool-definition-adapter.ts [hunk 3]
```diff
@@ -133,6 +136,7 @@ export function toToolDefinitions(tools: AnyAgentTool[]): ToolDefinition[] {
               toolName: name,
               params,
               toolCallId,
+              ctx: hookContext,
```
[the remark this answers is not in the document]
Comment (Reapor-Yurnero) [reply src/agents/pi-tool-definition-adapter.ts:139]: this is valid, but an existing upstream issue out of scope this PR

File: src/agents/pi-tools.before-tool-call.ts [hunk 4]
```diff
@@ -191,15 +219,16 @@ export async function runBeforeToolCallHook(args: {
...
-      toolContext,
-    );
+    const hookEvent = freezeHookSnapshot({
+      toolName,
+      params: cloneHookSnapshot(normalizedParams),
+      ...(args.ctx?.runId && { runId: args.ctx.runId }),
+      ...(args.toolCallId && { toolCallId: args.toolCallId }),
+      ...(args.ctx?.getSystemPrompt && { systemPrompt: args.ctx.getSystemPrompt() }),
+      ...(args.ctx?.getMessages && { messages: cloneHookSnapshot(args.ctx.getMessages()) }),
+      ...(args.ctx?.getTools && { tools: cloneHookSnapshot(args.ctx.getTools()) }),
```
[the remark this answers is not in the document]
Comment (Reapor-Yurnero) [reply src/agents/pi-tools.before-tool-call.ts:229]: not super critical as this has been the default behavior even for the pre-existing hook context. can address if needed.
